Refs #28643 -- Added math database functions.

JunyiJ opened the pull request:
Added ACOS, ASIN, ATAN, ATAN2, CEILING, COS, COT, DEGREES, EXP,
FLOOR, LOG, MOD, PI, POWER, RADIANS, ROUND, SIN, SQRT, TAN
ABS was added according to the commit by onkruid. 
Any suggestion is welcome!

JunyiJ commented:
Sorry that I did't use the up-to-date django master. As a result, there are conflicts. Could anyone help me to close this pull request?
Thanks

pope1ni commented:
@JunyiJ - This pull request is full of merge conflicts. Please can you resolve these first?

JunyiJ commented:
Hi I am working on it. Sorry that I am not so familiar with the whole process. I followed the instructions and ran tests locally.

loic commented:
At some point you must have pulled or rebase with master, and committed the conflict themselves (all the `<<<<<<< HEAD`, `=======`, `>>>>>>>` thingies).

atombrella commented on django/db/backends/sqlite3/base.py:
```diff
@@ -470,3 +497,233 @@ def _sqlite_regexp(re_pattern, re_string):
+    res = None
+    try:
+        res = x**y
+    except TypeError as e:
```
I think you could merge into one branch.

JunyiJ replied:
Hi atombrella,

Thank you so much for the reviews. I will work on it and try to finish it within one week. Sorry that I am a little busy this weekend and early next week.

atombrella commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,201 @@
+
+class Round(Transform):
+	function = 'ROUND'
+	lookup_name = 'Round'
```
Should be lowercase.

atombrella commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,201 @@
+    function = 'MOD'
+    arity = 2
+    output_field = fields.FloatField()
+    def __init__(self, *expressions, **extra):
```
Not needed due to `arity`.

atombrella commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,201 @@
+    function = 'PI'
+    arity = 0
+    output_field = fields.FloatField()
+    def __init__(self, *expressions, **extra):
```
Redundant.

atombrella commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,201 @@
+class Log(Transform):
+	function = 'LOG'
+	lookup_name = 'Log'
+	def as_postgresql(self, compiler, connection):
```
Please try to follow the indentation style. Same goes for blank lines between the function definitions.

JunyiJ replied:
Hi atombrella,
I just corrected the codes and reformat them as you suggested. Thank you very much. All failed tests have the same issue: pr/python/python3.5/tests/db_functions/test_math.py", line 5, in <module> from django.db.models.functions import (
ImportError: cannot import name 'PI'. However, I works just fine when I do the test locally.
Do you know what is the problem?
Best,
Junyi

JunyiJ commented:
I found that the error that causes all tests failure states that ```
pr/python/python3.5/tests/db_functions/test_math.py", line 5, in <module>
    from django.db.models.functions import (
ImportError: cannot import name 'PI'
```.
It seems that the test_math.py cannot import my math functions. However, it works just fine for me locally. Could anybody help me with that?

Thanks!
Junyi

loic commented on docs/ref/models/database-functions.txt:
```diff
@@ -680,6 +680,492 @@ that deal with time-parts can be used with ``TimeField``::
+        y = models.FloatField()
+
+``Abs``
+=========
```
Extra underlines characters. A few titles have this issue.

JunyiJ replied:
Thanks! Working on it.

loic commented:
> It seems that the test_math.py cannot import my math functions. However, it works just fine for me locally. Could anybody help me with that?

Are you still struggling with that?

It seems a bit odd to me, PY2 could have got mixed up between the absolute `math` package and the relative one, but PY3 doesn't have this issue.

Have you tried clearing `__pycache__`?

JunyiJ commented:
@loic I think I found the reason. Thank you very much.

JunyiJ commented:
@loic @atombrella Thank you for your suggestions and review. I fixed most of the problems. The only issue that I cannot resolve now is warnings generated by docs test. Any suggestions are appreciated. It says that:

```
/home/jenkins/workspace/docs/docs/releases/2.1.txt:176: WARNING: py:class reference target not found: django.db.models.functions.Abs
/home/jenkins/workspace/docs/docs/releases/2.1.txt:179: WARNING: py:class reference target not found: django.db.models.functions.Acos
/home/jenkins/workspace/docs/docs/releases/2.1.txt:182: WARNING: py:class reference target not found: django.db.models.functions.Asin
/home/jenkins/workspace/docs/docs/releases/2.1.txt:185: WARNING: py:class reference target not found: django.db.models.functions.Atan
/home/jenkins/workspace/docs/docs/releases/2.1.txt:188: WARNING: py:class reference target not found: django.db.models.functions.Atan2
/home/jenkins/workspace/docs/docs/releases/2.1.txt:191: WARNING: py:class reference target not found: django.db.models.functions.Ceiling
/home/jenkins/workspace/docs/docs/releases/2.1.txt:195: WARNING: py:class reference target not found: django.db.models.functions.Cos
/home/jenkins/workspace/docs/docs/releases/2.1.txt:198: WARNING: py:class reference target not found: django.db.models.functions.Cot
/home/jenkins/workspace/docs/docs/releases/2.1.txt:201: WARNING: py:class reference target not found: django.db.models.functions.Degrees
/home/jenkins/workspace/docs/docs/releases/2.1.txt:204: WARNING: py:class reference target not found: django.db.models.functions.Exp
/home/jenkins/workspace/docs/docs/releases/2.1.txt:208: WARNING: py:class reference target not found: django.db.models.functions.Floor
/home/jenkins/workspace/docs/docs/releases/2.1.txt:212: WARNING: py:class reference target not found: django.db.models.functions.Log
/home/jenkins/workspace/docs/docs/releases/2.1.txt:215: WARNING: py:class reference target not found: django.db.models.functions.Mod
/home/jenkins/workspace/docs/docs/releases/2.1.txt:218: WARNING: py:class reference target not found: django.db.models.functions.PI
/home/jenkins/workspace/docs/docs/releases/2.1.txt:221: WARNING: py:class reference target not found: django.db.models.functions.Power
/home/jenkins/workspace/docs/docs/releases/2.1.txt:224: WARNING: py:class reference target not found: django.db.models.functions.Radians
/home/jenkins/workspace/docs/docs/releases/2.1.txt:227: WARNING: py:class reference target not found: django.db.models.functions.Round
/home/jenkins/workspace/docs/docs/releases/2.1.txt:230: WARNING: py:class reference target not found: django.db.models.functions.Sin
/home/jenkins/workspace/docs/docs/releases/2.1.txt:233: WARNING: py:class reference target not found: django.db.models.functions.Sqrt
/home/jenkins/workspace/docs/docs/releases/2.1.txt:236: WARNING: py:class reference target not found: django.db.models.functions.Tan
/home/jenkins/workspace/docs/docs/releases/2.1.txt:239: WARNING: py:class reference target not found: django.db.models.DateFielarc cosined
```

atombrella commented on django/db/models/functions/__init__.py:
```diff
@@ -5,6 +5,10 @@
     TruncQuarter, TruncSecond, TruncTime, TruncWeek, TruncYear,
 )
+from .math import (
+    PI, Abs, Acos, Asin, Atan, Atan2, Ceiling, Cos, Cot, Degrees, Exp, Floor,
```
This should be sorted in alphabetical order. Same goes for the list in `__all__`.

JunyiJ replied:
@atombrella As you suggested, I changed the order. However, the isort test gave error message and suggested that I should put PI as the first import element. Should I follow what the isort result suggested?
Junyi

atombrella commented on docs/releases/2.1.txt:
```diff
@@ -173,8 +173,71 @@ Models
+
 * The new :class:`~django.db.models.functions.TruncWeek` function truncates
-  :class:`~django.db.models.DateField` and
+  :class:`~django.db.models.DateFielarc cosined` and
```
Not sure how you got this in here, but please undo this edit.

JunyiJ replied:
Thanks. I guess there was a mistake for copy and paste. I didn't notice that. I will undo it.

atombrella commented:
`.. module:: django.db.models.functions.math` remove this, and try again. The toplevel `..versionadded: 2.1` ought be enough. `PI` should be `Pi` IMO.

JunyiJ commented:
@atombrella Thank you very much for the suggestions! I followed your instructions and now it works.

JunyiJ commented:
@auvipy It supports sqlite, Mysql, postgresql and oracle.

pope1ni commented:
buildbot, test on oracle.

JunyiJ commented:
@pope1ni The test on oracle backend paused because of conflicts of releases/2.1.txt. It seems that the release/2.1.txt has been updated since my last commit. I am trying to resolve this conflict.

atombrella commented on django/db/backends/sqlite3/base.py:
```diff
@@ -476,3 +494,193 @@ def _sqlite_regexp(re_pattern, re_string):
+    res = None
+    try:
+        res = math.tan(x)
+    except TypeError as e:
```
You don't have tests for any of these exceptions. I'm not sure if the exceptions should pass through Django and be handled by the database. This is essentially the behavior of the three other backends.

atombrella commented on docs/ref/models/database-functions.txt:
```diff
@@ -982,3 +982,545 @@ gaps.
+
+.. class:: Pi(**extra)
+
+Accepts zero arguments and returns value of pi. 
```
`Returns the value of the mathematical constant Pi` sounds a bit better to me.

pope1ni replied:
Why not use the actual π symbol?

atombrella commented on docs/releases/2.1.txt:
```diff
@@ -155,7 +155,8 @@ Internationalization
 ~~~~~~~~~~~~~~~~~~~
 
-* ...
+* The new :option:`inspectdb --include-views` option allows creating models
```
Not sure why you get this change into the PR. Perhaps something about the tracking branch? Anyway, it shouldn't be here.

pope1ni replied:
This looks like an error in conflict resolution?

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -476,3 +494,193 @@ def _sqlite_regexp(re_pattern, re_string):
+    if x == 0:
+        return None
+    try:
+        res = -math.tan(3.0 * math.pi / 2.0 + x)
```
Cotangent is the reciprocal of tangent, so this should be `1 / math.tan(x)`.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -982,3 +982,545 @@ gaps.
+
+.. class:: Acos(expression, **extra)
+
+Accepts a single numeric field or expression and returns the value whose cosine is expression.Expression should be within the range -1 to 1.
```
Add missing space between sentences: `... expression. Expression ...`

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -476,3 +494,193 @@ def _sqlite_regexp(re_pattern, re_string):
+def _sqlite_mod(x, y):
+    res = None
+    try:
+        res = x % y
```
I suggest checking the behaviour of `MOD()` for other backends to ensure that the behaviour implemented here is as close as possible because, in Python, `x % y` differs to `math.fmod(x, y)`:

```python
>>> 5 % 3
2
>>> -5 % 3
1
>>> 5 % -3
-1
>>> -5 % -3
-2
>>> math.fmod(5, 3)
2.0
>>> math.fmod(-5, 3)
-2.0
>>> math.fmod(5, -3)
2.0
>>> math.fmod(-5, -3)
-2.0
```

Indeed, the documentation states:
```
>>> help(math.fmod)
Help on built-in function fmod in module math:

fmod(...)
    fmod(x, y)
    
    Return fmod(x, y), according to platform C.  x % y may differ.
```

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -476,3 +494,193 @@ def _sqlite_regexp(re_pattern, re_string):
+    return math.pi
+
+
+def _sqlite_power2(x, y):
```
With removal of the exception handling, remove this in favour of the existing `_sqlite_power()`.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -982,3 +982,545 @@ gaps.
+
+.. class:: Cot(expression, **extra)
+
+version added:: 2.1
```
(It was incorrect anyway - should have been `.. versionadded:: 2.1`)

pope1ni commented on django/db/models/functions/__init__.py:
```diff
@@ -5,6 +5,10 @@
     TruncQuarter, TruncSecond, TruncTime, TruncWeek, TruncYear,
 )
+from .math import (
+    Abs, Acos, Asin, Atan, Atan2, Ceiling, Cos, Cot, Degrees, Exp, Floor, Log,
```
In [#28643 (comment:9)](https://code.djangoproject.com/ticket/28643#comment:9) there are a list of recommended names for these new functions which is already being followed in other PRs. Pay particular attention to capitalisation.

Please can you change to use `ACos`, `ASin`, `ATan`, `ATan2` and `Ceil` throughout - code, documentation, etc.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,117 @@
+    output_field = fields.FloatField()
+
+
+class Ceiling(Transform):
```
In [#28643 (comment:9)](https://code.djangoproject.com/ticket/28643#comment:9) it recommends naming this function `Ceil` which feels more Pythonic as we have `math.ceil()`.

Please rename this function to `Ceil` and change `lookup_name` to `ceil`.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,117 @@
+    output_field = fields.FloatField()
+
+
+class Sqrt(Transform):
```
I think `Sqrt` needs `output_field = fields.FloatField()`.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,117 @@
+    output_field = fields.FloatField()
+
+
+class Round(Transform):
```
I haven't had time to look, but we should see what additional arguments `ROUND()` can take in different backends.

JunyiJ replied:
Indeed round also behaves differently in different database. For example, in mysql ROUND(X), ROUND(X,D): Rounds the argument X to D decimal places. The rounding algorithm depends on the data type of X. D defaults to 0 if not specified. 

In oracle ROUND returns n rounded to integer places to the right of the decimal point. If you omit integer, then n is rounded to 0 places.

felixxm replied:
IMO it works exactly the same way in [Oracle](https://docs.oracle.com/database/121/SQLRF/functions169.htm#SQLRF00698) and [MySQL](https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_round) database, am I missing sth? 🤔

atombrella replied:
Maybe it could be generalized by always passing the explicit argument `D` to 0, if not called differently.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,117 @@
+    lookup_name = 'floor'
+
+
+class Log(Transform):
```
I think that we should use `Log()` for `LOG(B, X)` and create `Ln()` for `LN(X)`. See the following for details:

```
PostgreSQL  MySQL           Oracle          SQLite
----------------------------------------------------------
LN(X)       LOG(X)/LN(X)    LN(x)            math.log(X)
---         LOG2(X)         ---              math.log2(X)
LOG(X)      LOG10(X)        ---              math.log10(X)
LOG(B, X)   LOG(B, X)       LOG(B, X)        math.log(X, B)

https://www.postgresql.org/docs/current/static/functions-math.html#FUNCTIONS-MATH-FUNC-TABLE
https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html
https://docs.oracle.com/en/database/oracle/oracle-database/12.2/sqlrf/LOG.html#GUID-3739F356-A4A0-4D0D-A4EB-9725ACA05CD1
https://docs.oracle.com/en/database/oracle/oracle-database/12.2/sqlrf/LN.html#GUID-DCC9EDAA-D308-4145-8E05-8D06A5EF5F6F
```

JunyiJ replied:
I have created Ln and LOG functions. I noticed that different database behaves slightly differently. For example, in postgresql the input for LOG(B, X) should be numeric but not double precision, meaning that the input shouldn't be float. As a result, I removed the test_log.py using float numbers as input. Another strange thing is that although spatialite also has a LOG(B, X) with the first parameter to be base, when I ran a test, the result seems to be the reciprocal to what I expected. As a result, I modified the test_log.py accordingly.

pope1ni commented:
(Sorry - GitHub seemed to break my review into pieces...)

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -179,10 +179,10 @@ def get_new_connection(self, conn_params):
+        conn.create_function("LOG", 2, _sqlite_log)
         conn.create_function("MOD", 2, _sqlite_mod)
         conn.create_function("PI", 0, _sqlite_pi)
-        conn.create_function("POWER", 2, _sqlite_power2)
```
I'd leave this in, but using `_sqlite_power`, even though `django_power` exists to avoid the need for `Power.as_sqlite()`.

timgraham commented on tests/db_functions/test_math.py:
```diff
@@ -0,0 +1,801 @@
+from .models import DecimalModel, FloatModel, IntegerModel
+
+
+class MathFUnctionTests(TestCase):
```
Yes, I think I'd do: `tests/db_functions/math/test_<func>.py`

JunyiJ commented:
@pope1ni @atombrella Thank you so much for the reviews and suggestions. I am working on them now. Thank you for the detailed explanations and the comments really helps a lot.

JunyiJ commented:
@pope1ni I have revised the code based on the review/comments and finished updating most of them. However, I noticed that there is no tests running for my last two commits. Is there any issue? Did I commit too much that the system prohibited running tests on my code?

timgraham commented:
You can squash your commits and force push. That should trigger a build.

JunyiJ commented:
@timgraham Thanks for the advice. I tried squash commits and force push, but it didn't trigger a build. Any suggestions?

timgraham commented on docs/releases/2.1.txt:
```diff
@@ -180,6 +180,70 @@ Models
 
 * Query expressions can now be negated using a minus sign.
 
+* The new :class:`~django.db.models.functions.Abs` database function returns
```
In this case, I think would be better to say something like "The new math database functions ... (some general)" rather than calling out every function here.

timgraham commented on tests/db_functions/test_sin.py:
```diff
@@ -0,0 +1,48 @@
+
```
chop blank line (check other files for this issue)

JunyiJ replied:
Thank you. I chopped the blank lines and updated releases/2.1.txt. The new commit still cannot trigger a build.

timgraham replied:
I saw an error in the Jenkins log and tried the steps here: https://stackoverflow.com/questions/36336188/jenkins2-branch-indexing-java-lang-illegalargumentexception-bytestring-null.

timgraham commented on docs/releases/2.1.txt:
```diff
@@ -180,69 +180,56 @@ Models
 
-* The new :class:`~django.db.models.functions.Abs` database function returns
-  the absolute value of an expression.
+* The new math database function Abs returns the absolute value of an
```
What I meant is to have only 1 bullet point and link "math database functions" to the appropriate section. Don't mention every function and what it does in the release notes.

JunyiJ replied:
I see. That's a good point. I have modified the doc accordingly. Now it starts to build after my most recent commit. Thank you!

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,117 @@
+class Acos(Transform):
+    function = 'ACOS'
+    lookup_name = 'acos'
+    output_field = fields.FloatField()
```
That's a good point, but I haven't find a good way yet. As you suggested, I documented it.

pope1ni replied:
Have you had any thoughts on this `FloatField` vs `DecimalField` issue?

atombrella commented on docs/ref/models/database-functions.txt:
```diff
@@ -680,6 +680,570 @@ that deal with time-parts can be used with ``TimeField``::
+Accepts a single numeric field or expression and returns the value whose cosine
+is expression. Expression should be within the range -1 to 1.
+
+.. warning::
```
PostgreSQL throws an error. I'm not sure about the other backends. Maybe this warning should be removed?

JunyiJ replied:
I removed it. Thanks.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+class ACos(Transform):
+    function = 'ACOS'
+    lookup_name = 'acos'
+    output_field = fields.FloatField()
```
Please import `FloatField` directly instead of the `fields` module.

JunyiJ replied:
Good point. Updated already. Thanks.

JunyiJ commented on django/db/backends/sqlite3/base.py:
```diff
@@ -469,3 +488,75 @@ def _sqlite_regexp(re_pattern, re_string):
+
+
+def _sqlite_acos(x):
+    return math.acos(x)
```
I have updated all of them. It is now more concise. Thanks!

JunyiJ commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,25 @@ def get_new_connection(self, conn_params):
+        conn.create_function("LOG", 2, _sqlite_log)
+        conn.create_function("MOD", 2, _sqlite_mod)
+        conn.create_function("PI", 0, _sqlite_pi)
+        conn.create_function("Power", 2, _sqlite_power)
```
Since there is a  _sqlite_power function in the same file already, I just use that function.

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,25 @@ def get_new_connection(self, conn_params):
+        conn.create_function('ACOS', 1, math.acos)
+        conn.create_function('ASIN', 1, math.asin)
+        conn.create_function('ATAN', 1, math.atan)
+        conn.create_function('ATAN2', 2, lambda x, y: math.atan2(x, y))
```
Sorry - my fault - don't need to wrap this one in a lambda.

pope1ni replied:
@JunyiJ - I think you missed this one.

JunyiJ replied:
Oops. Right. I have updated it. Thanks.

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,25 @@ def get_new_connection(self, conn_params):
+        conn.create_function('DEGREES', 1, math.degrees)
+        conn.create_function('EXP', 1, math.exp)
+        conn.create_function('FLOOR', 1, math.floor)
+        conn.create_function('ln', 1, lambda x: math.log(x))
```
Looks like this one got lowercased by mistake.

JunyiJ replied:
Updated. Thanks!

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,25 @@ def get_new_connection(self, conn_params):
+        conn.create_function('ln', 1, lambda x: math.log(x))
+        conn.create_function('LOG', 2, lambda x, y: math.log(y, x))
+        conn.create_function('MOD', 2, lambda x, y: math.fmod(x, y))
+        conn.create_function('PI', 0, _sqlite_pi)
```
This can be `lambda: math.pi`

JunyiJ replied:
Thanks!

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,122 @@
+from django.db.models import Func, Transform
+from django.db.models.fields import FloatField
```
`FloatField` can be imported from `django.db.models`, so merge with the line above.

JunyiJ replied:
Hi, I tried:
from django.db.models import Func, Transform, .fields.FloatField
But it doesn't work.

pope1ni replied:
I meant: `from django.db.models import FloatField, Func, Transform`

JunyiJ replied:
Thanks!

pope1ni commented on docs/releases/2.1.txt:
```diff
@@ -220,6 +220,72 @@ Models
+  the tangent of an expression.
+* The new math database functions Abs, ACos, ASin, ATan, ATan2, Ceil, Cos, Cot, 
+  Degrees, Exp, Floor, Ln, Log, Mod, Pi, Power, Radians, Round, Sin, Sqrt, Tan
+  are added and can be imported from :class:`~django.db.models.functions`.
```
Something has gone wrong here - I think that only this last bullet point should be included.

timgraham replied:
I think it's enough to say
```
Added many :ref:`math database functions <math-functions>`.
```

JunyiJ replied:
Thanks, I have changed it.

pope1ni replied:
I think you have forgotten to push the change!

JunyiJ replied:
Sorry. My fault. I have updated and pushed it.

timgraham commented on tests/db_functions/test_abs.py:
```diff
@@ -0,0 +1,42 @@
+
+class AbsTests(TestCase):
+
+    def test_abs_decimal(self):
```
Please omit "_abs" and similar in all test names.

JunyiJ replied:
Thanks. I have put all tests into a new folder tests/db_functions/math.

timgraham commented on tests/db_functions/test_abs.py:
```diff
@@ -0,0 +1,42 @@
+
+    def test_abs_transform(self):
+        try:
+            DecimalField.register_lookup(Abs, 'abs')
```
The second argument to register_lookup() and _unregister_lookup() should be omitted as it will default to `lookup_name`.

JunyiJ replied:
Thanks!

timgraham commented on tests/db_functions/test_sin.py:
```diff
@@ -0,0 +1,43 @@
+        obj = IntegerModel.objects.annotate(
+            small_d=Sin('small'),
+            normal_d=Sin('normal'),
+            big_d=Sin('big')).first()
```
In cases like this, `).first()` should go on the next line.

JunyiJ replied:
Thanks. Updated.

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,27 @@ def get_new_connection(self, conn_params):
+        conn.create_function('DEGREES', 1, math.degrees)
+        conn.create_function('EXP', 1, math.exp)
+        conn.create_function('FLOOR', 1, math.floor)
+        conn.create_function('LN', 1, lambda x: math.log(x))
```
The lambda wrapper is not need here.

JunyiJ replied:
I have updated the line 
conn.create_function('LN', 1, lambda x: math.log(x))
to
conn.create_function('LN', 1, math.log)

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -169,6 +169,27 @@ def get_new_connection(self, conn_params):
+        conn.create_function('FLOOR', 1, math.floor)
+        conn.create_function('LN', 1, lambda x: math.log(x))
+        conn.create_function('LOG', 2, lambda x, y: math.log(y, x))
+        conn.create_function('MOD', 2, lambda x, y: math.fmod(x, y))
```
The lambda wrapper is not need here.

JunyiJ replied:
I have updated the line
conn.create_function('LN', 1, lambda x: math.log(x))
to
conn.create_function('LN', 1, math.log)

I think the functions with two arguments should keep the lambda wrapper. Correct me if I was wrong. Thanks!

pope1ni replied:
I was referring to `lambda x, y: math.fmod(x, y)` where the arguments are passed in the same order.

JunyiJ replied:
I see. I have updated the conn.create_function('MOD', ...) and conn.create_function('ATAN2', ...) since atan2 also has arguments in the same order.

JunyiJ commented:
Hi @pope1ni I have resolved the conflicts now. Thank you. The conflicts resolving part is more complicated than I expected.

pope1ni commented on django/db/backends/sqlite3/base.py:
```diff
@@ -171,6 +171,29 @@ def get_new_connection(self, conn_params):
+        conn.create_function('ACOS', 1, math.acos)
+        conn.create_function('ASIN', 1, math.asin)
+        conn.create_function('ATAN', 1, math.atan)
+        conn.create_function('ATAN2', 2, math.atan2)
```
Nearly... You've duplicated `ACOS`, `ASIN`, `ATAN` and `ATAN2`.

timgraham commented on tests/db_functions/math/test_log.py:
```diff
@@ -0,0 +1,33 @@
+    def test_decimal(self):
+        DecimalModel.objects.create(n1=Decimal('12.9'), n2=Decimal('3.6'))
+        obj = DecimalModel.objects.annotate(n_log=Log('n1', 'n2')).first()
+        self.assertTrue(
```
This looks difficult to debug.  If it fails all we'll see is something like "False is not True". Is there any way to formulate it using other assertions? If not, maybe add a comment about it? Adding an assertion helper naming something like `assertLogEqual` could reduce repetition.

JunyiJ replied:
Thank you for the suggestion! I added a helper function assertLogEqual as you suggested.

timgraham commented on docs/spelling_wordlist:
```diff
@@ -2,6 +2,7 @@
 accessor
 accessors
 Aceh
+ACos
```
I don't think these should be needed. The functions should always be in backticks prevents them from being spell checked.

JunyiJ replied:
OK. I removed all the function names I added.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -680,6 +680,565 @@ that deal with time-parts can be used with ``TimeField``::
+ 
+    >>> from django.db.models import FloatField
+    >>> from django.db.models.functions import Abs
+    >>> FloatField.register_lookup(Abs, 'abs')
```
Similar to the tests, the second argument isn't needed, correct?

JunyiJ replied:
Right. I have corrected all of the test cases.

felixxm commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class Pi(Func):
+    function = 'PI'
```
`PI` doesn't exist on Oracle. It is a constant, therefore we can use `math.pi`.

JunyiJ replied:
@felixxm, please check this link
https://docs.oracle.com/javadb/10.8.3.0/ref/rreffuncpi.html

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class Degrees(Transform):
+    function = 'DEGREES'
```
HI @felixxm, I found this page https://docs.oracle.com/javadb/10.8.3.0/ref/rreffuncdegrees.html
which is the degrees function on Oracle. However, I am not an Oracle user and please correct me if I was wrong.

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class Ceil(Transform):
+    function = 'CEILING'
```
Hi @felixxm I am wondering what are the difference between CEIL and CEILING? I didn't find the difference based on the link: https://docs.oracle.com/javadb/10.8.3.0/ref/rreffuncceil.html

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class Cot(Transform):
+    function = 'COT'
```
Hi @felixxm, I found the COT function here: https://docs.oracle.com/javadb/10.8.3.0/ref/rreffunccot.html

atombrella replied:
The documentation is for a different product. The structure of `docs.oracle.com` can be a bit confusing as it's for a lot of products. You can use https://docs.oracle.com/cd/E11882_01/index.htm

felixxm replied:
@JunyiJ you refer to Java DB documentation. In Oracle DB you can use only functions from [this](https://docs.oracle.com/database/121/SQLRF/functions002.htm#SQLRF51178) list.

JunyiJ replied:
@felixxm Thank you very much! I will update it soon!

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class Radians(Transform):
+    function = 'RADIANS'
```
Hi @felixxm, please check this link: https://docs.oracle.com/javadb/10.8.3.0/ref/rreffuncradians.html

atombrella commented on tests/db_functions/math/test_acos.py:
```diff
@@ -0,0 +1,44 @@
+        obj = IntegerModel.objects.annotate(
+            small_acos=ACos('small'),
+            normal_acos=ACos('normal'),
+            big_acos=ACos('big')
```
Please add a trailing comma.

JunyiJ commented:
I am a little bit surprised that none of the automatic tests test database oracle (correct me if I am wrong). @felixxm @pope1ni @atombrella @timgraham

JunyiJ commented on tests/db_functions/math/test_tan.py:
```diff
@@ -0,0 +1,43 @@
+        obj = IntegerModel.objects.annotate(
+            small_d=Tan('small'),
+            normal_d=Tan('normal'),
+            big_d=Tan('big')).first()
```
All cases have been updated.

JunyiJ commented on docs/ref/models/database-functions.txt:
```diff
@@ -680,6 +680,565 @@ that deal with time-parts can be used with ``TimeField``::
+
+Returns the arc tangent of expression1 and expression2.
+
+.. warning::
```
Removed all unnecessary warning.

timgraham commented:
buildbot, test on oracle.

JunyiJ commented:
buildbot, test on oracle.

JunyiJ commented:
buildbot, test on oracle.

JunyiJ commented:
buildbot, test on oracle.

JunyiJ commented:
builtbot, test on oracle.

JunyiJ commented:
buildbot test on oracle

felixxm commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,171 @@
+
+    def as_oracle(self, compiler, connection):
+        cot = 1 / self.get_tan()
+        return float(cot)
```
@JunyiJ My previous suggestion was to use the `TAN` database function on Oracle, i.e. 
```python
def as_oracle(self, compiler, connection):
    return super().as_sql(compiler, connection, template='1 / TAN(%(expressions)s)')
```

JunyiJ replied:
I see. Thanks @felixxm. I have to admit that I am not so familiar with the uses of template and as_sqal(...). I updated the Cot already. Do you mind to help me check the other oracle functions: Degrees, Pi and Radians in the directory: django/db/models/functions/math.py ? Testing on oracle is too slow (generally 1 hour to run the oracle test).

felixxm commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,167 @@
+    output_field = FloatField()
+
+    def __init__(self, num):
+        super().__init__(num)
```
`Cot.__init__()` is unnecessary.

JunyiJ commented:
buildbot, test on oracle.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,158 @@
+        expressions = [self.get_degrees()]
+        clone = self.copy()
+        clone.set_source_expressions(expressions)
+        return super(Degrees, clone).as_sql(compiler, connection)
```
This is clunky - the extra function, assignments, unnecessary casting to `float()`.

Regardless, this will not work. What if the source expression is `Value(5.5)` or some other complex expression?

I think this is what you are looking for:

```python
def as_oracle(self, compiler, connection):
    return super().as_sql(compiler, connection, template='((%%(expressions)s) * %s / 180)' % math.pi)
```

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,158 @@
+        expressions = [self.get_pi()]
+        clone = self.copy()
+        clone.set_source_expressions(expressions)
+        return super(Pi, clone).as_sql(compiler, connection)
```
```python
def as_oracle(self, compiler, connection):
    return super().as_sql(compiler, connection, template=str(math.pi))
```

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,158 @@
+        expressions = [self.get_radians()]
+        clone = self.copy()
+        clone.set_source_expressions(expressions)
+        return super(Radians, clone).as_sql(compiler, connection)
```
```python
def as_oracle(self, compiler, connection):
    return super().as_sql(compiler, connection, template='((%%(expressions)s) * 180 / %s)' % math.pi)
```

JunyiJ replied:
Thanks! I think the degrees and radians should be switched. I have handled it.

pope1ni replied:
Argh. Thanks :man_facepalming:

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,158 @@
+    output_field = FloatField()
+
+    def as_oracle(self, compiler, connection):
+        return super().as_sql(compiler, connection, template='1 / TAN(%(expressions)s)')
```
Should the content of the template be wrapped in parentheses? (To avoid issues when functions are nested...)

What do you think, @felixxm?

felixxm commented:
@pope1ni Totally agreed with all your comments, thanks! :bowing_man: :rocket:

JunyiJ commented:
Thank you @pope1ni @felixxm for helping the oracle functions. I have committed all the changes as you suggested.

felixxm commented:
buildbot, test on oracle.

JunyiJ commented:
Hi @pope1ni @felixxm @timgraham Thank you so much for your comments and review suggestions. I am wondering whether there is anything else to change or not? If not, could you help test on oracle and then merge it?

pope1ni commented:
Thank you for your patience, @JunyiJ. This is coming together well now.

There are a couple of things that I can think of outstanding.

1. As suggested by @felixxm, we can now remove `django_power`:
    ```diff
    diff --git a/django/db/backends/sqlite3/base.py b/django/db/backends/sqlite3/base.py
    index a9753bb094..07b153cd22 100644
    --- a/django/db/backends/sqlite3/base.py
    +++ b/django/db/backends/sqlite3/base.py
    @@ -6,2 +6,3 @@ import decimal
     import math
    +import operator
     import re
    @@ -170,3 +171,2 @@ class DatabaseWrapper(BaseDatabaseWrapper):
             conn.create_function("django_format_dtdelta", 3, _sqlite_format_dtdelta)
    -        conn.create_function("django_power", 2, _sqlite_power)
             conn.create_function('LPAD', 3, _sqlite_lpad)
    @@ -187,3 +187,3 @@ class DatabaseWrapper(BaseDatabaseWrapper):
             conn.create_function('PI', 0, lambda: math.pi)
    -        conn.create_function('POWER', 2, _sqlite_power)
    +        conn.create_function('POWER', 2, operator.pow)
             conn.create_function('RADIANS', 1, math.radians)
    @@ -498,5 +498 @@ def _sqlite_rpad(text, length, fill_text):
         return (text + fill_text * length)[:length]
    -
    -
    -def _sqlite_power(x, y):
    -    return x ** y
    diff --git a/django/db/backends/sqlite3/operations.py b/django/db/backends/sqlite3/operations.py
    index 10b064d966..65796ea8a2 100644
    --- a/django/db/backends/sqlite3/operations.py
    +++ b/django/db/backends/sqlite3/operations.py
    @@ -274,6 +274,6 @@ class DatabaseOperations(BaseDatabaseOperations):
         def combine_expression(self, connector, sub_expressions):
    -        # SQLite doesn't have a power function, so we fake it with a
    -        # user-defined function django_power that's registered in connect().
    +        # SQLite doesn't have a POWER function, so we fake it with a
    +        # user-defined function that's registered in connect().
             if connector == '^':
    -            return 'django_power(%s)' % ','.join(sub_expressions)
    +            return 'POWER(%s)' % ','.join(sub_expressions)
             return super().combine_expression(connector, sub_expressions)
    ```

2. I raised the issue of input values being `DecimalField` and coming out as `FloatField`.

    The functions `ACos`, `ASin`, `ATan`, `ATan2`, `Cos`, `Cot`, `Degrees`, `Log`, `Mod`, `Pi`, `Power`, `Radians`, `Sin`, `Sqrt` and `Tan` all declare `output_field = FloatField()`.

    Firstly, this doesn't seem consistent as some functions are missing this, e.g. `Ln`. It makes sense that `IntegerField` becomes `FloatField`, but I'd expect `DecimalField` to stay `DecimalField`.

    One way this could be achieved would be via a hack similar to this:
   https://github.com/django/django/blob/d549b8805053d4b064bf492ba90e90db5d7e2a6b/django/db/models/functions/window.py#L49-L51
    But as a mixin so that it can be added easily to each of the math functions that require it:
    ```python
    class MathOutputFieldMixin:
        def _resolve_output_field(self):
            sources = self.get_source_expressions()
            if any(isinstance(s.output_field, DecimalField) for s in sources):
                return DecimalField()
            else:
                return super()._resolve_output_field(self)
    ```
    I'd like some opinion on this from @felixxm and @timgraham before you spend time implementing it.

pope1ni commented on tests/db_functions/math/test_cos.py:
```diff
@@ -0,0 +1,44 @@
+        ).first()
+        self.assertAlmostEqual(float(obj.small_cos), math.cos(obj.small))
+        self.assertAlmostEqual(float(obj.normal_cos), math.cos(obj.normal))
+        self.assertAlmostEqual(float(obj.big_cos), math.cos(obj.big))
```
Remove the cast to `float` for these - it shouldn't be necessary due to `output_field = FloatField()`.

pope1ni commented on tests/db_functions/math/test_cot.py:
```diff
@@ -0,0 +1,44 @@
+        ).first()
+        self.assertAlmostEqual(float(obj.small_cot), 1 / math.tan(obj.small))
+        self.assertAlmostEqual(float(obj.normal_cot), 1 / math.tan(obj.normal))
+        self.assertAlmostEqual(float(obj.big_cot), 1 / math.tan(obj.big))
```
Remove the cast to `float` for these - it shouldn't be necessary due to `output_field = FloatField()`.

pope1ni commented on tests/db_functions/math/test_atan.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('3.12'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('-5'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__atan__gt=0)
+            self.assertQuerysetEqual(objs, [float(3.12)], lambda a: float(a.n1))
```
The type casting here is weird - `float(3.12)` is pointless as it can be written `3.12`.
I suspect that you shouldn't need `float` in the lambda either as you didn't do it for `ASin`...

JunyiJ replied:
I have to say it is weird. The reason I wrote float(3.12) is that is I didn't add it, the `[3.12]` gives the result `[Decimal(3.12]` and the lambda function gives result `[3.12]`. I am wondering why `[3.12]` gives the result `[Decimal(3.12]`.

JunyiJ replied:
I found it will work if I change it to `self.assertQuerysetEqual(objs, [3.12], lambda a: float(a.n1))`. It seems that the type of 3.12 is somehow related to the return type of the lambda function.

pope1ni commented on tests/db_functions/math/test_ceil.py:
```diff
@@ -0,0 +1,38 @@
+            DecimalModel.objects.create(n1=Decimal('3.12'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('1.25'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__ceil__gt=3)
+            self.assertQuerysetEqual(objs, [float(3.12)], lambda a: float(a.n1))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_degrees.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Degrees('f1'), f2_d=Degrees('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.degrees(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.degrees(obj.f2))
```
Remove the cast to `float` for these - it shouldn't be necessary due to `output_field = FloatField()`.

pope1ni commented on tests/db_functions/math/test_degrees.py:
```diff
@@ -0,0 +1,44 @@
+        ).first()
+        self.assertAlmostEqual(float(obj.small_d), math.degrees(obj.small))
+        self.assertAlmostEqual(float(obj.normal_d), math.degrees(obj.normal))
+        self.assertAlmostEqual(float(obj.big_d), math.degrees(obj.big))
```
Remove the cast to `float` for these - it shouldn't be necessary due to `output_field = FloatField()`.

pope1ni commented on tests/db_functions/math/test_degrees.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('-30'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__degrees__gt=0)
+            self.assertQuerysetEqual(objs, [float(5.4)], lambda a: float(a.n1))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_exp.py:
```diff
@@ -0,0 +1,44 @@
+        ).first()
+        self.assertEqual(obj.small_d, int(math.exp(obj.small)))
+        self.assertEqual(obj.normal_d, int(math.exp(obj.normal)))
+        self.assertEqual(obj.big_d, int(math.exp(obj.big)))
```
Adding `output_field = FloatField()` to `Exp` will mean that this doesn't need casting to `int` which is wrong.

pope1ni commented on tests/db_functions/math/test_exp.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Exp('f1'), f2_d=Exp('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.exp(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.exp(obj.f2))
```
With the values already being `float` in `FloatModel`, the casting shouldn't be required.

pope1ni commented on tests/db_functions/math/test_exp.py:
```diff
@@ -0,0 +1,44 @@
+        DecimalModel.objects.create(n1=Decimal('-12.9'), n2=Decimal('0.6'))
+        obj = DecimalModel.objects.annotate(n1_d=Exp('n1'), n2_d=Exp('n2')).first()
+        self.assertAlmostEqual(float(obj.n1_d), math.exp(obj.n1))
+        self.assertAlmostEqual(float(obj.n2_d), math.exp(obj.n2))
```
Adding `output_field = FloatField()` to `Exp` will mean that this doesn't need casting to `float`.

pope1ni commented on tests/db_functions/math/test_floor.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Floor('f1'), f2_d=Floor('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.floor(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.floor(obj.f2))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_floor.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('3.4'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__floor__gt=4)
+            self.assertQuerysetEqual(objs, [float(5.4)], lambda a: float(a.n1))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_ln.py:
```diff
@@ -0,0 +1,44 @@
+        ).first()
+        self.assertEqual(obj.small_d, int(math.log(obj.small)))
+        self.assertEqual(obj.normal_d, int(math.log(obj.normal)))
+        self.assertEqual(obj.big_d, int(math.log(obj.big)))
```
Adding `output_field = FloatField()` to `Ln` will mean that this doesn't need casting to `int` which is wrong.

pope1ni commented on tests/db_functions/math/test_ln.py:
```diff
@@ -0,0 +1,44 @@
+        DecimalModel.objects.create(n1=Decimal('12.9'), n2=Decimal('0.6'))
+        obj = DecimalModel.objects.annotate(n1_d=Ln('n1'), n2_d=Ln('n2')).first()
+        self.assertAlmostEqual(float(obj.n1_d), math.log(obj.n1))
+        self.assertAlmostEqual(float(obj.n2_d), math.log(obj.n2))
```
Adding `output_field = FloatField()` to `Ln` will mean that this doesn't need casting to `float`.

pope1ni commented on tests/db_functions/math/test_ln.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Ln('f1'), f2_d=Ln('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.log(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.log(obj.f2))
```
With the values already being `float` in `FloatModel`, the casting shouldn't be required.

pope1ni commented on tests/db_functions/math/test_log.py:
```diff
@@ -0,0 +1,34 @@
+from ..models import DecimalModel, IntegerModel
+
+
+class LogTests(TestCase):
```
Why does this differ so much to the other tests? What is `assertLogEqual()` doing that isn't required by `LnTests`? Is this a hang over from before `assertAlmostEqual()` was being used and it wasn't correctly updated?

Also missing `test_float()`.

JunyiJ replied:
Thanks, I have corrected it.

Thinking back, I might (wrongly) thought that some database use LOG(BASE, NUM) and some use LOG(NUM, BASE). However, today when I double checked, I found that all database use the form LOG(BASE, NUM) and only python use LOG(NUM, BASE).

pope1ni commented on tests/db_functions/math/test_log.py:
```diff
@@ -0,0 +1,34 @@
+    def assertLogEqual(self, n_log, n1, n2):
+        diff1 = abs(float(n_log) - math.log(n2, n1))
+        diff2 = abs(float(n_log) - math.log(n1, n2))
+        if(diff1 < diff2):
```
Drop parentheses and add a space after `if`: `if diff1 < diff2:`

This feels like it was trying to work around parameters being passed in the wrong order...

pope1ni commented on tests/db_functions/math/test_mod.py:
```diff
@@ -0,0 +1,20 @@
+from ..models import IntegerModel
+
+
+class ModTests(TestCase):
```
Missing `test_decimal()` and `test_float()`.

pope1ni commented on tests/db_functions/math/test_pi.py:
```diff
@@ -0,0 +1,14 @@
+    def test(self):
+        FloatModel.objects.create(f1=2.5, f2=15.9)
+        obj = FloatModel.objects.annotate(pi=Pi()).first()
+        self.assertAlmostEqual(obj.pi, math.pi, places=5)
```
Is this not accurate to more than 5 decimal places on all backends? It is not that important, I am just surprised.

JunyiJ replied:
I cannot remember whether I placed it on purpose or not. I removed it to give a try.

pope1ni commented on tests/db_functions/math/test_power.py:
```diff
@@ -0,0 +1,30 @@
+        ).first()
+        self.assertAlmostEqual(obj.small_d, obj.small**obj.normal)
+        self.assertAlmostEqual(obj.normal_d, obj.normal**obj.big)
+        self.assertAlmostEqual(obj.big_d, obj.big**obj.small)
```
Add a space either side of `**`.

pope1ni commented on tests/db_functions/math/test_power.py:
```diff
@@ -0,0 +1,30 @@
+    def test_float(self):
+        FloatModel.objects.create(f1=2.3, f2=1.1)
+        obj = FloatModel.objects.annotate(f1_d=Power('f1', 'f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), obj.f1**obj.f2)
```
Add a space either side of `**`, casting to `float` should not be necessary.

pope1ni commented on tests/db_functions/math/test_power.py:
```diff
@@ -0,0 +1,30 @@
+    def test_decimal(self):
+        DecimalModel.objects.create(n1=Decimal('1.0'), n2=Decimal('-0.6'))
+        obj = DecimalModel.objects.annotate(n1_d=Power('n1', 'n2')).first()
+        self.assertAlmostEqual(obj.n1_d, obj.n1**obj.n2)
```
Add a space either side of `**`.

pope1ni commented on tests/db_functions/math/test_radians.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Radians('f1'), f2_d=Radians('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.radians(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.radians(obj.f2))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_sin.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Sin('f1'), f2_d=Sin('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.sin(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.sin(obj.f2))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_sin.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('0.1'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__sin__lt=0)
+            self.assertQuerysetEqual(objs, [float(5.4)], lambda a: float(a.n1))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_sqrt.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Sqrt('f1'), f2_d=Sqrt('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.sqrt(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.sqrt(obj.f2))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_tan.py:
```diff
@@ -0,0 +1,44 @@
+        FloatModel.objects.create(f1=-27.5, f2=0.33)
+        obj = FloatModel.objects.annotate(f1_d=Tan('f1'), f2_d=Tan('f2')).first()
+        self.assertAlmostEqual(float(obj.f1_d), math.tan(obj.f1))
+        self.assertAlmostEqual(float(obj.f2_d), math.tan(obj.f2))
```
Again, the `float` type casting should be unnecessary.

pope1ni commented on tests/db_functions/math/test_ceil.py:
```diff
@@ -0,0 +1,38 @@
+        FloatModel.objects.create(f1=-12.5, f2=21.33)
+        obj = FloatModel.objects.annotate(f1_ceil=Ceil('f1'), f2_ceil=Ceil('f2')).first()
+        self.assertAlmostEqual(int(obj.f1_ceil), int(math.ceil(obj.f1)))
+        self.assertAlmostEqual(int(obj.f2_ceil), int(math.ceil(obj.f2)))
```
It shouldn't be necessary to cast these to `int`.

JunyiJ commented:
Hi @pope1ni, I revised the code based on your suggestions except the output_field(Though I think it is a great idea to use output_field_reolved, I will wait to see whether there is different opinion). There are several issues that I noticed:
1) For the log(a,b)function, different database behaves differently. Some use log(base, num) and some use log(num, base).
2) Some data base doesn't support mod(double precision, double precision). I guess that is the reason why I didn't include the tests on float for the mod function.
3) Some data base doesn't support log(double precision, double precision). I guess that is the reason why I didn't include the tests on float for the mod function.

For cases 2) and 3) Do you think it is OK to remove the tests  on float?

Thanks,
Junyi

pope1ni commented on tests/db_functions/math/test_atan.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('3.12'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('-5'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__atan__gt=0)
+            self.assertQuerysetEqual(objs, [3.12], lambda a: float(a.n1))
```
Simplify to `lambda a: a.n1`

JunyiJ replied:
That's gonna give me error message `AssertionError: Lists differ: [Decimal('3.12')] != [3.12]`

JunyiJ replied:
Instead, I changed it to `self.assertQuerysetEqual(objs, [Decimal('3.12')], lambda a: a.n1)`.
Problem solved.

pope1ni replied:
Yes. That's what it should be.

pope1ni commented on tests/db_functions/math/test_ceil.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('3.12'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('1.25'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__ceil__gt=3)
+            self.assertQuerysetEqual(objs, [3.12], lambda a: float(a.n1))
```
Simplify to `lambda a: a.n1`

pope1ni commented on tests/db_functions/math/test_degrees.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('-30'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__degrees__gt=0)
+            self.assertQuerysetEqual(objs, [5.4], lambda a: float(a.n1))
```
Simplify to `lambda a: a.n1`

pope1ni commented on tests/db_functions/math/test_floor.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('3.4'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__floor__gt=4)
+            self.assertQuerysetEqual(objs, [5.4], lambda x: float(x.n1))
```
Simplify to `lambda a: a.n1` (also reverting `x` to `a`)

pope1ni commented on tests/db_functions/math/test_sin.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('0.1'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__sin__lt=0)
+            self.assertQuerysetEqual(objs, [5.4], lambda a: float(a.n1))
```
Simplify to `lambda a: a.n1`

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,169 @@
+import math
+
+from django.db.models import FloatField, Func, Transform
+from .comparison import Cast
```
`from django.db.models.functions import Cast`

pope1ni commented on tests/db_functions/math/test_floor.py:
```diff
@@ -0,0 +1,44 @@
+            DecimalModel.objects.create(n1=Decimal('5.4'), n2=Decimal('0'))
+            DecimalModel.objects.create(n1=Decimal('3.4'), n2=Decimal('0'))
+            objs = DecimalModel.objects.filter(n1__floor__gt=4)
+            self.assertQuerysetEqual(objs, [Decimal('5.4')], lambda x: x.n1)
```
Please change this back to `lambda a: a.n1` for consistency with all of the other tests.

pope1ni commented on tests/db_functions/math/test_log.py:
```diff
@@ -0,0 +1,39 @@
+
+class LogTests(TestCase):
+
+    def assertLogEqual(self, n_log, n1, n2):
```
Please remove this again and change back to `assertAlmostEqual`.

I know that it causes a test failure at the moment, but we need to fix it properly. This assertion method just makes the assumption that passing in one way or the other works. We have already ascertained that all backends that define the function use `LOG(B, X)`, and Python has `math.log(x, b)` which is already catered for in the function creation for `sqlite`.

The issue here is that `sqlite` is working and `spatialite` is failing with the arguments in the wrong order. This is strange as `spatialite` either ought to use the same function as `sqlite` or what looks like it's own function which ought to be compatible: [`LOG(B, X)`](http://www.gaia-gis.it/gaia-sins/spatialite-sql-4.3.0.html#math).

pope1ni commented on tests/db_functions/math/test_ceil.py:
```diff
@@ -0,0 +1,44 @@
+        try:
+            DecimalField.register_lookup(Ceil)
+            DecimalModel.objects.create(n1=3.12, n2=0)
+            DecimalModel.objects.create(n1=1.25, n2=0)
```
Please revert to using `Decimal` here:
```python
            DecimalModel.objects.create(n1=Decimal('3.12'), n2=Decimal('0'))
            DecimalModel.objects.create(n1=Decimal('1.25'), n2=Decimal('0'))
```

JunyiJ commented on django/db/models/functions/math.py:
```diff
@@ -77,19 +78,58 @@ class Floor(Transform):
+            return super(Log, clone).as_sql(
+                compiler, connection, function='LOG', template='%(function)s(%(expressions)s)'
+            )
+        return self.as_sql(compiler, connection)
```
Looks much better than my old version. However, after I changed it there is error message:
`psycopg2.DataError: invalid input syntax for integer: "none"
LINE 1: ...odel"."f2", LOG(("db_functions_floatmodel"."f1")::numeric(No...`
@pope1ni

pope1ni commented:
Hi @JunyiJ,

I've made a number of fixes that you can pull into your branch @ https://github.com/JunyiJ/django/pull/1:

- Simplified the type casting for `Log()` and `Mod()` on PostgreSQL.
  *(The issue was that a value is needed for `max_digits` and `decimal_places`.)*
- We no longer cast `DecimalField` to `FloatField` for `output_field`.
- Added type assertions to tests to complement the `output_field` stuff.
- Added missing tests for `Round()`.
- Fixed the problem with `Log()` on the SpatiaLite backend.
- Various other bits of tidying in the tests.

I've noted that `ATAN2` was added to SpatiaLite v4.3.0 which also seems to exhibit problems but I think Django's Jenkins currently uses < v4.3.0, so `math.atan2` is still being used.

See http://www.gaia-gis.it/gaia-sins/spatialite-sql-4.3.0.html#math

Bug filed for Log() on SpatiaLite: https://www.gaia-gis.it/fossil/libspatialite/tktview?name=8f59ddebf0

JunyiJ commented:
@pope1ni Thank you very much for fixing these problems. I just merged all your changes to the branch.

Junyi

pope1ni commented:
buildbot, test on oracle.

pope1ni commented:
So apparently it looks like spatialite handles `ROUND()` differently to everyone else too.

Just checking that Oracle is behaving...

JunyiJ commented:
@pope1ni In that case, we can either loose the test case criteria to pass the tests or change the Round function in math.py as what you did to Log. What do you think is a better way?

pope1ni commented:
> In that case, we can either loose the test case criteria to pass the tests...

No, we shouldn't do this. The behaviour of spatialite is incorrect. Tests should test correct behaviour, not be made lax to satisfy broken implementations. Could you revert that change?

I'll do some more digging over the next couple of days and try and come up with a correct solution. I have a few ideas.

JunyiJ commented:
@pope1ni OK. I have reverted the implementation. 

Based on the definition SpatiaLite round function returns the integer value nearest to x, but it doesn't specify the special case of numbers ended with 0.5. As a result, I am not sure whether SpatiaLite Round function is wrong here or it is designed to be so.

wkschwartz commented on django/db/backends/sqlite3/base.py:
```diff
@@ -168,9 +169,27 @@ def get_new_connection(self, conn_params):
+        conn.create_function('ATAN2', 2, math.atan2)
+        conn.create_function('CEILING', 1, math.ceil)
+        conn.create_function('COS', 1, math.cos)
+        conn.create_function('COT', 1, lambda x: 1 / math.tan(x))
```
I haven’t verified this myself, but [this SO answer](https://stackoverflow.com/a/3738696/1286628) says it’s better to use cos/sin than 1/tan. Intuitively I understand it near pi/2 but I would expect the opposite to be true close to zero. Not sure if there’s a one best option here. Just wanted to raise this in case anyone had opinions.

wkschwartz replied:
Maybe one option is to split the inputs based on whether they’re closer to +/-pi/2 or +/- pi (mod 2pi).

pope1ni replied:
@wkschwartz: Is there a better source for this recommendation to use `lambda x: math.cos(x) / math.sin(x)`?

If we were to make this change, we should add a comment here to explain why.

timgraham commented:
Nick, do you want to look at this any more per https://github.com/django/django/pull/9622#issuecomment-376322325?

Junyi, please target Django 2.2 and squash the commits.

pope1ni commented:
Tim, will aim to have a look at this over the next few days.

JunyiJ commented:
@timgraham Hi Tim, Does 'Targeting Django 2.2' means that I should update the 2.2 in documents folder? 
@pope1ni Hi Nick, should I wait until you finish reviewing?

Thanks!
Junyi

pope1ni commented:
So https://github.com/django/django/pull/9622#issuecomment-376322325 was already addressed by Junyi's last commit - it seems that `ROUND()` is unstable around `0.5`.

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,164 @@
+    lookup_name = 'cot'
+
+    def as_oracle(self, compiler, connection):
+        return super().as_sql(compiler, connection, template='(1 / TAN(%(expressions)s))')
```
Following on from the previous comment, if we were to make the change, this should also be changed accordingly.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -680,6 +680,485 @@ that deal with time-parts can be used with ``TimeField``::
+Math Functions
+==============
+
+.. versionadded:: 2.1
```
Please update all of these to `2.2`

pope1ni commented on docs/releases/2.1.txt:
```diff
@@ -223,6 +223,8 @@ Models
 * :meth:`.QuerySet.order_by` and :meth:`distinct(*fields) <.QuerySet.distinct>`
   now support using field transforms.
 
+* Added many :ref:`math database functions <math-functions>`.
```
Please move this to `docs/releases/2.2/txt`

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,164 @@
+    lookup_name = 'atan'
+
+
+class ATan2(OutputFieldMixin, Func):
```
Support was added for `ATan2()` in SpatiaLite 4.3.0, but it fails the tests.

```
======================================================================
FAIL: test_decimal (db_functions.math.test_atan2.ATan2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/nick/Sources/django/tests/db_functions/math/test_atan2.py", line 16, in test_decimal
    self.assertAlmostEqual(obj.n_atan2, Decimal(math.atan2(obj.n1, obj.n2)))
  File "/usr/lib/python3.6/unittest/case.py", line 878, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: Decimal('2.70662570254421') != Decimal('-1.135829375749316749732997777755372226238250732421875') within 7 places

======================================================================
FAIL: test_float (db_functions.math.test_atan2.ATan2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/nick/Sources/django/tests/db_functions/math/test_atan2.py", line 22, in test_float
    self.assertAlmostEqual(obj.f_atan2, math.atan2(obj.f1, obj.f2))
  File "/usr/lib/python3.6/unittest/case.py", line 878, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: 3.128393420165654 != -1.5575970933707572 within 7 places
```
Strangely, after inverting the order of the arguments - similar to the issue with `Log()` - the float test passes and the integer test passes when casting the inputs to floats.

So the following method needs to be added:
```python
    def as_sqlite(self, compiler, connection):
        if not getattr(connection.ops, 'spatialite', False):
            return self.as_sql(compiler, connection)
        if connection.ops.spatial_version < (4, 3, 0):
            return self.as_sql(compiler, connection)
        # As documented this function ought to be ATan2(y, x) returning the
        # inverse tangent of y / x, but it turns out that, unlike all other
        # backends, the function is actually ATan2(x, y). In addition, it
        # doesn't handle integer inputs properly, so these need to be casted to
        # floats.
        # Although the spatialite backend inherits from the sqlite3 backend,
        # the spatialite function takes precedence over the one created using
        # create_function().
        clone = self.copy()
        clone.set_source_expressions([
            Cast(expression, FloatField()) if not isinstance(expression.output_field, FloatField)
            else expression for expression in self.get_source_expressions()[::-1]
        ])
        return clone.as_sql(compiler, connection)
```

pope1ni replied:
Sorry - I should clarify - the tests were failing for float and decimal.
This was fixed by inverting the order of the arguments, but that caused integer to fail.
Casting the arguments to float seemed to then make the integer test pass.

If somebody else could scrutinise my suggested fix, that would be great.

JunyiJ replied:
@pope1ni Thank you for the instructions and detailed analysis. It is weird that the Atan2 function cannot pass SpatiaLite tests. Should we consider to use as_sql(...) for all SpatiaLite Atan2 functions?

pope1ni replied:
I'm not sure why the change I suggested to add `ATan2.as_sqlite()` targeting SpatiaLite >= 4.3.0 resolved this issue for me locally and not on Jenkins.

Could you confirm the version of SpatiaLite on the xenial and bionic boxes, @timgraham?

(Maybe a build just needs to be re-triggered.)

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,184 @@
+        if not getattr(connection.ops, 'spatialite', False):
+            return self.as_sql(compiler, connection)
+        if connection.ops.spatial_version < (4, 3, 0):
+            return self.as_sql(compiler, connection)
```
Sorry - I could have done this more cleanly:
```python
        if not getattr(connection.ops, 'spatialite', False) or connection.ops.spatial_version < (4, 3, 0):
            return self.as_sql(compiler, connection)
```

pope1ni commented on django/db/models/functions/math.py:
```diff
@@ -0,0 +1,184 @@
+        # create_function().
+        clone = self.copy()
+        clone.set_source_expressions([
+            Cast(expression, FloatField()) if not isinstance(expression.output_field, FloatField)
```
Could you also change this line for me?
```diff
-            Cast(expression, FloatField()) if not isinstance(expression.output_field, FloatField)
+            Cast(expression, FloatField()) if isinstance(expression.output_field, IntegerField)
```
Don't forget to import `IntegerField`.

pope1ni commented:
buildbot, test on oracle.

pope1ni commented:
So I have found the cause and am now happy with the workaround.

Bug filed for ATan2() on SpatiaLite: https://www.gaia-gis.it/fossil/libspatialite/tktview?name=0f72cca3a2

JunyiJ commented:
Thanks, @pope1ni. I have updated atan2 based on your suggestions. The only issue left (I hope) seems to the error when tested on oracle, saying that `django.db.utils.ProgrammingError: positional and named binds cannot be intermixed`. I don't think I changed anything related to the backend base.py for oracle.

timgraham commented:
The Oracle failure is [a separate issue](https://code.djangoproject.com/ticket/29541).

timgraham commented:
I pushed some edits. Let me know if there are any outstanding issues.

pope1ni commented:
Thanks for polishing the docs. Just a few tweaks and I think we're good.

Were you planning to land this before #10121 and then have that updated to add `@none_guard` to the functions that need it?

adamchainz commented:
🎉 Thanks everyone for the hard work  🎉
